File: src/1038.binary-search-tree-to-greater-sum-tree.1114/solution.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
var bstToGst = function (root) {
    // 遍历求和
    const sum = getSum(root);
    // BST 中序遍历 按从小到大遍历
    root && inOrder(root, sum);
    return root;
};

function inOrder (node, sum) {
    if (node.left) {
        sum = inOrder(node.left, sum);
    }
    const tmp = node.val;
    node.val = sum;
    sum -= tmp;
    if (node.right) {
        sum = inOrder(node.right, sum);
    }

    return sum;
}

function getSum (node) {
    if (!node) {
        return 0;
    }

    return node.val + getSum(node.left) + getSum(node.right);
}

File: src/1038.binary-search-tree-to-greater-sum-tree.1114/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int sum = 0;
    public TreeNode bstToGst(TreeNode root) {
        if(root == null){
            return null;
        }
        bstToGst(root.right);
        
        sum += root.val;
        root.val = sum;

        bstToGst(root.left);
        
        return root;
    }
    
}